project/Mima.scala

```diff
@@ -35,6 +35,8 @@ object Mima {
     ProblemFilters.exclude[A]("scala.meta." + metaType)
 
   val apiCompatibilityExceptions: Seq[ProblemFilter] = Seq(
+    // Placeholder
+    exclude[IncompatibleMethTypeProblem]("Type#Placeholder.unapply"),
```

Comment:
@bjaglin earlier, `unapply` took a `Placeholder` parameter whereas now it's a `Type` instead. this makes it a source-compatible change but perhaps not a binary-compatible one.

Reply:
Good point. It does look like https://github.com/xuwei-k/scalafix-rules/blob/v0.2.8/rules/src/main/scala/fix/Scala3Placeholder.scala would fail at runtime against newer versions of scalafix :disappointed: 

```
➜  javap -p -c -cp $(cs fetch com.github.xuwei-k::scalafix-rules:0.2.8 | head -1) 'fix.Scala3Placeholder$$anonfun$fix$1' | grep Placeholder\$\.unapply
      23: invokevirtual #49                 // Method scala/meta/Type$Placeholder$.unapply:(Lscala/meta/Type$Placeholder;)Lscala/Option;
      20: invokevirtual #49                 // Method scala/meta/Type$Placeholder$.unapply:(Lscala/meta/Type$Placeholder;)Lscala/Option;
```

What about adding back that overload for binary compatibility? Would scalac handle it well and discard it in favor of the wider overload when pattern matching on a tree?

Reply:
i tried this. it complained about unapply overloads. what else can be tried?

Reply:
Interestingly, it seems that the [error](https://github.com/scala/scala/blob/d978429e66b0c93ee804883abd693d5651c17eef/src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala#L324) is only raised when pattern matching is used from synthetics, not from source. Not sure why...

https://github.com/scalameta/scalameta/blob/a4319baa16618bd848bbcffbb328a56ba1aa362c/scalameta/common/shared/src/main/scala/scala/meta/internal/transversers/transverser.scala#L64 could be rewritten to use the right overload explicitly I guess? But that would mean clients pattern matching on `Placeholder` from macros would get the same problem.

I see you are now exploring the subclassing option :+1: Eventually removing `Placeholder` from the hierarchy will impact binary compatibility for `AnonymousParam` & `Wildcard`, but that can't happen until 5.x anyway, so binary compatibility is best-effort.